### PR TITLE
Ajouter la recherche dans les données des établissements

### LINKED ISSUE #84: Ajouter les établissements dans la recherche

### PR DESCRIPTION
**Pull request by HAEKADI**
closes #84

### CODE REVIEW

**Review by XavierJp — APPROVED** [review]

File: aio/aio-proxy/aio_proxy/search/filters/boolean.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
+from aio_proxy.search.helpers.elastic_fields import get_elasticsearch_field_name
+from elasticsearch_dsl import Q
+
+
+def filter_search_by_bool_fields_unite_legale(
+    search, filters_to_include: list, **params
+):
+    for param_name, param_value in params.items():
```
Comment (XavierJp) [inline aio/aio-proxy/aio_proxy/search/filters/boolean.py:8]: Maybe more explicits tests :

```
isParamDefined = param_value is not None and param_name in filters_to_include
```

File: aio/aio-proxy/aio_proxy/search/filters/nested_etablissements_filters.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    text_filters = ["departement", "code_postal", "commune"]
+
+    terms_filters = []
+    must_filters = []
+    must_not_filters = []
+
+    # params is the list of parameters (filters) provided in the request
+    for param_name, param_value in params.items():
+
+        if param_value is not None and param_name in text_filters:
```
Comment (XavierJp) [inline aio/aio-proxy/aio_proxy/search/filters/nested_etablissements_filters.py:43]: idem would benefit from an explicit variable

File: aio/aio-proxy/aio_proxy/search/filters/nested_etablissements_filters.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+def build_nested_etablissements_filters_query_with_inner_hits(**params):
+
+    filters_query = {
+        "nested": {
+            "path": "etablissements",
+            "query": {"bool": {}},
+            "inner_hits": {},
+        }
+    }
+
```
Comment (XavierJp) [inline aio/aio-proxy/aio_proxy/search/filters/nested_etablissements_filters.py:87]: Maybe try to merge `build_nested_etablissements_filters_query_with_inner_hits`
`build_nested_etablissements_filters_query_without_inner_hits`

into 
`build_nested_etablissements_filters_query`

with an optional param `add_inner_hits`

```
   filters_query = {
        "nested": {
            "path": "etablissements",
            "query": {"bool": {}},
        }
    }
   if (add_inner_hits) :
       filters_query.nested['inner_hits'] = {}
```

File: aio/aio-proxy/aio_proxy/search/parsers/siren.py [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+
+def is_siren(query_string: str) -> bool:
+    """
+    Check if string is siren (composed of 9 digits).
+    """
+    if query_string is None:
+        return False
+    clean_query_string = query_string.replace(" ", "").upper()
+    siren_valides = r"\b\d{9}\b"
```
Comment (XavierJp) [inline aio/aio-proxy/aio_proxy/search/parsers/siren.py:11]: .upper() seems useless here as siren is composed with digits

File: aio/aio-proxy/aio_proxy/search/parsers/siret.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import re
+
+
+def is_siret(query_string: str) -> bool:
+    """
+    Check if string is siret (composed of 14 digits).
+    """
+    if query_string is None:
+        return False
+    clean_query_string = query_string.replace(" ", "").upper()
```
Comment (XavierJp) [inline aio/aio-proxy/aio_proxy/search/parsers/siret.py:10]: iDDem

File: aio/aio-proxy/aio_proxy/search/queries/text.py [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+                        "query": {
+                            "match_phrase": {
+                                "identifiant_association_unite_legale": {
+                                    "query": terms,
+                                    "boost": 50,
+                                    "_name": "exact match identifiant association",
+                                }
+                            }
+                        },
+                        "field_value_factor": {
```
Comment (XavierJp) [inline aio/aio-proxy/aio_proxy/search/queries/text.py:16]: this code is used a lot. Maybe use a variable ?

```
etablissement_ouvert_multiplier = {
                            "field": "nombre_etablissements_ouverts",
                            "factor": 1,
                            "modifier": "log2p",
                            "missing": 1,
                        }
```

And then everywhere you need it 

```
"field_value_factor": etablissement_ouvert_multiplier
```

File: aio/aio-proxy/aio_proxy/search/text_search.py [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+
+
+def text_search(index, offset: int, page_size: int, **params):
+    query_terms = params["terms"]
+    s = index.search()
+
+    # Check if any etablissements filters are used
+    etablissement_filter_used = is_any_etablissement_filter_used(**params)
+
+    # Filter by siren first (if query is a `siren` number), and return search results
```
Comment (XavierJp) [inline aio/aio-proxy/aio_proxy/search/text_search.py:28]: You can factoriz a bit as siren and siret really follow the same logic

```
is_siren  = 
is_siret = 
if(is_siren or is_siret): 
   query_terms_clean = query_terms.replace(" ", "")
   s = is_siren ? filter_by_siren(s, query_terms_clean) : filter_by_sirent(s, query_terms_clean)

        return sort_and_execute_search(
            search=s,
            offset=offset,
            page_size=page_size,
            is_search_fields=False,
        )
```

File: aio/aio-proxy/aio_proxy/search/text_search.py [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+from aio_proxy.search.queries.person import search_person
+from aio_proxy.search.queries.text import build_text_query
+from elasticsearch_dsl import Q
+
+
+def text_search(index, offset: int, page_size: int, **params):
+    query_terms = params["terms"]
+    s = index.search()
+
+    # Check if any etablissements filters are used
```
Comment (XavierJp) [inline aio/aio-proxy/aio_proxy/search/text_search.py:25]: Why check this before siren / siret ?

File: aio/aio-proxy/aio_proxy/search/text_search.py [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+    # include inner hits in this particular search query
+    if etablissement_filter_used:
+        filters_etablissements_query_without_inner_hits = (
+            build_nested_etablissements_filters_query_without_inner_hits(**params)
+        )
+        if filters_etablissements_query_without_inner_hits:
+            s = s.query(Q(filters_etablissements_query_without_inner_hits))
+
+    # Filters applied on établissements without text search
+    if not query_terms and etablissement_filter_used:
```
Comment (XavierJp) [inline aio/aio-proxy/aio_proxy/search/text_search.py:83]: indent and remove last  `etablissement_filter_used` ?

File: aio/aio-proxy/aio_proxy/search/text_search.py [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+            text_query, **params
+        )
+        s = s.query(Q(text_query_with_filters))
+
+    # Text search only without etablissements filters
+    if query_terms and not etablissement_filter_used:
+        text_query = build_text_query(query_terms)
+        s = s.query(Q(text_query))
+
+    is_search_fields = False
```
Comment (XavierJp) [inline aio/aio-proxy/aio_proxy/search/text_search.py:175]: Maybe use a more explicit name `is_textual_search` ?

File: aio/aio-proxy/aio_proxy/search/text_search.py [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+                    "type_person": "colter_elus",
+                    "match_nom": "nom",
+                    "match_prenom": "prenom",
+                    "match_date": "date_naissance",
+                },
+            ],
+            **params,
+        )
+
+    # Filters applied on établissement with text search
```
Comment (XavierJp) [inline aio/aio-proxy/aio_proxy/search/text_search.py:162]: Move up, indent and remove `etablissement_filter_used`

File: aio/aio-proxy/aio_proxy/search/text_search.py [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+    # Boolean filters for unité légale
+    s = filter_search_by_bool_fields_unite_legale(
+        s,
+        filters_to_include=[
+            "est_association",
+            "est_collectivite_territoriale",
+        ],
+        **params,
+    )
+
```
Comment (XavierJp) [inline aio/aio-proxy/aio_proxy/search/text_search.py:99]:
```suggestion
types = [colter_elus, dirigeant_pp]

if params["type_personne"] and params["type_personne"] in [ELU,DIRIGEANT]:
  types = [params["type_personne"]]
  
s = search_person(
            s,
            "nom_personne",
            "prenoms_personne",
            "min_date_naiss_personne",
            "max_date_naiss_personne",
            [
                [{
                    "type_person": type_person,
                    "match_nom": "nom",
                    "match_prenom": "prenom",
                    "match_date": "date_naissance",
                } for type_person in types]
            ],
            **params,
        )
```

**Review by XavierJp — CHANGES_REQUESTED** [review]

File: aio/aio-proxy/aio_proxy/response/format_response.py [hunk 13]
```diff
@@ -93,14 +70,26 @@ def get_field(field, default=None):
                 "est_ess": format_ess(
                     get_field("economie_sociale_solidaire_unite_legale")
                 ),
-                "est_finess": format_bool_field(get_field("liste_finess")),
-                "est_rge": format_bool_field(get_field("liste_rge")),
-                "est_uai": format_bool_field(get_field("liste_uai")),
+                "est_finess": format_etablissements_list(get_field("etablissements"))[
```
[the remark this answers is not in the document]
Comment (HAEKADI) [reply aio/aio-proxy/aio_proxy/response/format_response.py]: I added the `complements` directly in elasticsearch.

File: aio/aio-proxy/aio_proxy/response/format_response.py [hunk 14]
```diff
@@ -93,14 +70,26 @@ def get_field(field, default=None):
...
                     "identifiant_association_unite_legale"
                 ),
             },
         }
+
+        # If 'etablissements' is included in elasticsearch response
+        etablissements = format_etablissements_list(get_field("etablissements"))[0]
+        if etablissements:
+            result_formatted["etablissements"] = etablissements
+
```
[the remark this answers is not in the document]
Comment (HAEKADI) [reply aio/aio-proxy/aio_proxy/response/format_response.py]: I added a var to deal with this issue.

File: aio/aio-proxy/aio_proxy/search/filters/boolean.py [hunk 15]
```diff
@@ -0,0 +1,21 @@
+from aio_proxy.search.helpers.elastic_fields import get_elasticsearch_field_name
+from elasticsearch_dsl import Q
+
+
```
[the remark this answers is not in the document]
Comment (HAEKADI) [reply aio/aio-proxy/aio_proxy/search/filters/boolean.py:4]: I think I will open another PR next week where I will document everything in more detail.

File: aio/aio-proxy/aio_proxy/search/queries/text.py [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+        "bool": {
+            "should": [
+                {
+                    "function_score": {
+                        "query": {
+                            "match_phrase": {
+                                "identifiant_association_unite_legale": {
+                                    "query": terms,
+                                    "boost": 50,
+                                    "_name": "exact match identifiant association",
```
Comment (geoffreyaldebert) [inline aio/aio-proxy/aio_proxy/search/queries/text.py:19]: cool the possibility to name each steps

File: aio/aio-proxy/aio_proxy/search/queries/text.py [hunk 17]
```diff
@@ -0,0 +1,213 @@
...
+                            "match": {
+                                "liste_elus": {
+                                    "query": terms,
+                                    "operator": "AND",
+                                    "boost": 10,
+                                    "_name": "partial match liste élus",
+                                }
+                            }
+                        },
+                        "field_value_factor": etablissements_ouverts_multiplier,
```
Comment (geoffreyaldebert) [inline aio/aio-proxy/aio_proxy/search/queries/text.py:207]: j'ai tout compris

**Review by geoffreyaldebert — APPROVED** [review]
congrats for this huge work 👏

**Comment by geoffreyaldebert** [comment]
but please lint this code !!

File: aio/aio-proxy/aio_proxy/response/format_search_results.py [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+                "est_uai": get_field("est_uai"),
+                "identifiant_association": get_field(
+                    "identifiant_association_unite_legale"
+                ),
+            },
+        }
+
+        # If 'include_etablissements' param is True, return 'etablissements' object
+        # even if it's empty, otherwise do not return object
+        etablissements = format_etablissements_list(get_field("etablissements"))
```
Comment (XavierJp) [inline aio/aio-proxy/aio_proxy/response/format_search_results.py:79]: you could only declare variable when `include_etablissements` is true, why declare it before the if statement ?

File: aio/aio-proxy/aio_proxy/response/formatters/etablissements.py [hunk 19]
```diff
@@ -0,0 +1,70 @@
+from aio_proxy.response.helpers import get_value
+
+
```
Comment (XavierJp) [inline aio/aio-proxy/aio_proxy/response/formatters/etablissements.py:3]: Very good 👌

File: aio/aio-proxy/aio_proxy/response/helpers.py [hunk 20]
```diff
@@ -1,4 +1,16 @@
 import json
+import os
+from dotenv import load_dotenv
+
+load_dotenv()
+
+current_env = os.getenv("ENV")
+
+
+def is_dev_env():
```
Comment (XavierJp) [inline aio/aio-proxy/aio_proxy/response/helpers.py:10]:
```suggestion
def is_dev_env():
  return current_env == "dev"
```

File: aio/aio-proxy/aio_proxy/search/filters/boolean.py [hunk 21]
```diff
@@ -0,0 +1,25 @@
...
+def filter_search_by_bool_fields_unite_legale(
+    search, filters_to_include: list, **params
+):
+    for param_name, param_value in params.items():
+        is_param_defined_and_a_bool_filter = (
```
Comment (XavierJp) [inline aio/aio-proxy/aio_proxy/search/filters/boolean.py:9]: Not sure what `and_a_bool_filter` means

File: aio/aio-proxy/aio_proxy/search/filters/nested_etablissements_filters.py [hunk 22]
```diff
@@ -0,0 +1,102 @@
...
+    id_filters = ["id_finess", "id_rge", "id_uai", "id_convention_collective"]
+    text_filters = ["departement", "code_postal", "commune"]
+
+    terms_filters = []
+    must_filters = []
+
+    # params is the list of parameters (filters) provided in the request
+    for param_name, param_value in params.items():
+        is_param_defined_and_a_text_filter = (
```
Comment (XavierJp) [inline aio/aio-proxy/aio_proxy/search/filters/nested_etablissements_filters.py:35]: Now I understand. I would simplify variable name.

eg `is_text_param_defined` or `is_text_filter`

Comment (XavierJp) [reply aio/aio-proxy/aio_proxy/search/filters/nested_etablissements_filters.py:35]: `should_apply_filter`

File: aio/aio-proxy/aio_proxy/search/helpers/color.py [hunk 23]
```diff
@@ -0,0 +1,31 @@
+import json
+import logging
+import os
+from urllib.request import urlopen
+
+from dotenv import load_dotenv
+
+load_dotenv()
```
Comment (XavierJp) [inline aio/aio-proxy/aio_proxy/search/helpers/color.py:8]: `is_env_dev` ?

File: aio/aio-proxy/aio_proxy/search/text_search.py [hunk 24]
```diff
@@ -0,0 +1,183 @@
...
+                )
+    else:
+        # Text search only without etablissements filters
+        if query_terms:
+            text_query = build_text_query(query_terms)
+            search_client = search_client.query(Q(text_query))
+
+    # Search 'élus' only or 'dirigeants' only
+    types_personnes = {"ELU": "colter_elus", "DIRIGEANT": "dirigeants_pp"}
+
```
Comment (XavierJp) [inline aio/aio-proxy/aio_proxy/search/text_search.py:118]:
```suggestion
type = params["type_personne"]

if type and type = 'ELUS':
  types = ['colter_elus']
elif type and type = 'DIRIGEANT':
  types = ['dirigeants_pp']
else:
  types = ['colter_elus','dirigeants_pp']
  
search_client = search_person(
            search_client,
            "nom_personne",
            "prenoms_personne",
            "min_date_naiss_personne",
            "max_date_naiss_personne",
            [
                {
                    "type_person": type,
                    "match_nom": "nom",
                    "match_prenom": "prenom",
                    "match_date": "date_naissance",
                } for type in types
            ],
            **params,
        )
```

**Review by XavierJp — APPROVED** [review]

File: aio/aio-proxy/aio_proxy/search/helpers/color.py (hunk 23 above)
Comment (HAEKADI) [reply aio/aio-proxy/aio_proxy/search/helpers/color.py]: it's another var that I use 'color_url` from env variables, but I use the same file to load it.

File: aio/aio-proxy/aio_proxy/search/text_search.py (hunk 24 above)
Comment (HAEKADI) [reply aio/aio-proxy/aio_proxy/search/text_search.py]: hee no du coup lol

**Pull request merged**